### ISSUE TITLE
Graph scalability: health filtering

### ISSUE THREAD

**Issue by lucasponce** [kiali/kiali#3252]
Looking at this graph:
![image](https://user-images.githubusercontent.com/1662329/94151474-bedf6c80-fe7a-11ea-8b65-37185eb32ca0.png)

I wonder if there is some possibility to add some "pre-defined" "find/hide" actions in a 1-click.

On this case I've found very useful to have some button to just focus on the unhealthy elements (and zoom them).

**Comment by jshaughn** [comment]
Find/Hide on Health seems like a needed extension as health becomes a greater focus.  I think we could do something relatively easy on Nodes, probably a unary operation like "is healthy" or "! healthy".

**Comment by lucasponce** [comment]
> probably a unary operation like "is healthy" or "! healthy".

But also some of this could be like buttons at the right of the find/hide toolbar

![image](https://user-images.githubusercontent.com/1662329/94578486-136a5980-0278-11eb-82da-b45197cb3ec2.png)

For most used operations we could enable "less typing" "find/hide" features.

I mean, I'd like to see the find/hide populated but also applied in the graph when using the shortcut button.

**Comment by jshaughn** [comment]
I'd prefer not to dedicate buttons for expressions and defer to the dropdown for previously entered expressions.   Too many buttons make the toolbar confusing and we don't really know which expressions a user really wants to use.  He may want to combine "!healthy" with AND or OR for example.  I'd rather we leave it to the user's history.  Or, perhaps look again at building a "favorites" feature using local storage.

**Comment by jshaughn** [comment]
Looking at the code I may have to make a slight change to the way we store health of the graph element, moving it from a cy "class" to another data element.  Classes are only good for "exists" type checks and are mainly used for display options.  They are not good for negative tests.

**Comment by lucasponce** [comment]
> I'd prefer not to dedicate buttons for expressions and defer to the dropdown for previously entered expressions. Too many buttons make the toolbar confusing and we don't really know which expressions a user really wants to use. He may want to combine "!healthy" with AND or OR for example.

Don't expect that as a user I know how to type that or I read a +10 lines help to do that.
There should be a toolbar with most common / used scenarios plus find/hide for custom/powerful/free expressions.

**Comment by jshaughn** [comment]
My point is that we may not know exactly what those common scenarios are.  And the graph toolbar is already dense.  I think some community feedback would be required before making this kind of an addition.

**Comment by jshaughn** [comment]
Hmmm, now I remember why I didn't initially add health to graph find/hide; health is determined asynchronously.  Async health setting is bad for find/hide because find executes when the graph refreshes, prior to health being applied to the nodes.  It is usually very fast but it's not guaranteed.  To have reasonable behavior we'd need to solve this problem, a solution is not obvious.

**Comment by lucasponce** [comment]
> My point is that we may not know exactly what those common scenarios are. And the graph toolbar is already dense. I think some community feedback would be required before making this kind of an addition.

So, the feedback your team is providing you is not valid ?

As a user, I see the find/hide as an "advanced" mode not suitable for users but I would appreciate to have shortcuts or easy ways to introduce common patterns.

**Comment by jshaughn** [comment]
> So, the feedback your team is providing you is not valid ?

Of course your opinion is valid, as valid as mine :-)

> I would appreciate to have shortcuts

We do have shortcuts in the form of the browser dropdown of suggestions.  If we were to add anything I'd suggest replacing the browser dropdown with maybe some sort of favorites, where the user can choose what's actually important.

The whole discussion about having find/hide on health is moot unless we solve the issue I mentioned above, because at the moment its not possible to perform the hide/find on healh.

**Comment by jshaughn** [comment]
OK, we had some off-issue discussion.  The idea of applying health as part of graph-gen on the server-side is not viable.  The reason is that the server-side gathers health information but final health status is calculated on the client-side.  So, in order to get find/hide to work with health the health assignments will need to be applied as part of the graph-gen, but also client-side.

Although this requires multiple round-trips to the server before we can render the graph it probably makes sense.  Today we render the graph prior to having the health applied, and in general no one notices, but if the lag for health was noticeable it could cause "flashing" in the graph, where nodes render and then change color.  So, a slight delay in graph rendering probably won't be noticeable but by waiting for health to be applied a more seemles experience should be guaranteed, as well as having the ability to offer health in find/hide.

I'll look at this approach and see if there are unoreseen issues, or whether it's viable.

**Comment by jotak** [comment]
@jshaughn given the new parts that are being included in health (such as traces) I think it's more important than ever to have it be fetched asynchronously without impacting the reactivity of other elements.
What about having the find/hide filters being re-rendered every time a health is fetched, but without *waiting* for it at the first pass? This re-rendering could be a little smart, e.g. trigger it only if it's detected that it depends on health ?

**Comment by jshaughn** [comment]
@jotak I agree that there could be some danger going forward but I think we need to find a different solution than a staggered re-rendering.  I think that approach could offer a bad experience to the user as they could see the graph flash/change/re-render while passively viewing, or worse, after having manually moved a node.  Also, I think it may be technically challenging/messy, knowing how things are architected wrt graph fetch and find/hide.

If health calculation gets very slow I think we will have a variety of problems.  One possible approach would be to be able to calculate "quick" or partial health, and then full-health in different, more drill-down or passive scenarios.  For example, the graph could use just traffic health, or maybe traffic and whatever else is deemed "fast".

**Comment by jshaughn** [comment]
@lucasponce @jotak I have the new approach working.  The good news is that health filtering is nice and we no longer have flashing of unhealthy nodes on refresh.  The bad news is, as we suspected, the amount of time added to graph rendering is high. Fetching health is accounting for 35-50% of the total time, and more usually ~45% when displaying service nodes (the default).  So, the render time is nearly double as health-generation takes almost as long as the graph-generation.

In terms of user experience, it may or may not be noticeable.  An initial fetch of travel+bookinfo takes about a second in my env. Subsequent fetch takes the same time but is almost imperceptible given the lack of flashing and assuming no real topology change.

I'm going to create a PR for people to try things out and we can decide how to proceed.  We could try and play some games, possibly adding health loading as part of graph state and using it to defer find/hide when necessary.  Or we could just release the new approach upstream and see if there is any negative reaction.  Either way we should start to keep an eye on health determination performance, as it is already fairly expensive.

**Comment by jshaughn** [comment]
> Or, if we assume that most if not all apps/workloads/services will have traffic, we could just request all app/workload/service health for the selected namespaces simultaneously with the graph-generation (no app health if graph type doesn't deal with apps).

UPDATE: This approach may be promising.  Although we may fetch more health than necessary we can do it simultaneously.  Some early tests look pretty good.  But there is a caveat, "outsider" nodes are in the graph but not in the requested namespaces.  We may need to get their health post-fetch.

**Comment by jshaughn** [comment]
@lucasponce @jotak @israel-hdez I think perhaps we should meet to discuss our options here, now and going forward.  Fundamentally we may need to reconcile what we're doing with health and how that plays with what we do with the graph.  To me there are two major issues that will affect our scale, first is that the traffic processing for graph and health is almost a duplication of effort.  They do things a bit differently but from the prometheus standpoint the load increases.  The difficulty here is that the processing is time-period based and typically rolling forward, so the results are dynamic.  The second is the non-traffic health, which likely does not change often but that we probably calculate repeatedly.

Also, there is the fact that we calculate global-health client-side, in the UI.  This currently inhibits us from calculating and/or stashing health server-side, possibly in a passive, cached way on the server.

**Comment by jotak** [comment]
> @lucasponce @jotak @israel-hdez I think perhaps we should meet to discuss our options here, now and going forward. 

good for me

Fundamentally we may need to reconcile what we're doing with health and how that plays with what we do with the graph. To me there are two major issues that will affect our scale, first is that the traffic processing for graph and health is almost a duplication of effort. They do things a bit differently but from the prometheus standpoint the load increases. The difficulty here is that the processing is time-period based and typically rolling forward, so the results are dynamic.

Yes, there's a big overlap here. Which means there's certainly some place for optimization. No need to compute error rates in health if it's already done in the graph. Perhaps we could add flags into the health API to tell which parts we want to fetch or not?

The second is the non-traffic health, which likely does not change often but that we probably calculate repeatedly.

But that's also the one that may grow in the future. I guess all the pods-related health (available vs desired, etc.) is cached via informers ... it should be double-checked.

> 
> Also, there is the fact that we calculate global-health client-side, in the UI. This currently inhibits us from calculating and/or stashing health server-side, possibly in a passive, cached way on the server.

And on the flip side, if we amend the health API as suggested above to not fetch error-rate health, it's on the UI that the health aggregation will happen, I guess?

**Comment by lucasponce** [comment]
Reading all comments I think we are mixing implementation details (how graph is rendered, how find/hide is implemented) with also design/architect topics (split requests for graph and health, client side calculation of health) and I'm sure in that way we can miss something relevant.

As commented I'll fetch both PRs to run some deep testing on my side.

> Or we could just release the new approach upstream and see if there is any negative reaction. Either way we should start to keep an eye on health determination performance, as it is already fairly expensive.

About this I'm ok to put this feature as "do not merge" yet if we have concerns about performance and/or how it can impact to the user experience.

**Comment by lucasponce** [comment]
Some backend considerations during PR review https://github.com/kiali/kiali-ui/pull/1957#issuecomment-704796442.

If you debug the Prom client, you'd see a pattern like this:

```
I1007 11:13:11.305598  479093 client.go:92] TODELETE GetAllRequestRates [alpha] [60s] [2020-10-07 11:13:11.305107932 +0200 CEST m=+18.023489238]
I1007 11:13:11.305641  479093 client.go:92] TODELETE GetAllRequestRates [alpha] [60s] [2020-10-07 11:13:11.305021299 +0200 CEST m=+18.023402559]
I1007 11:13:11.307653  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [alpha] [60s] [2020-10-07 11:13:11.307551725 +0200 CEST m=+18.025933005]
I1007 11:13:11.308251  479093 client.go:92] TODELETE GetAllRequestRates [beta] [60s] [2020-10-07 11:13:11.30767096 +0200 CEST m=+18.026052216]
I1007 11:13:11.309466  479093 client.go:92] TODELETE GetAllRequestRates [beta] [60s] [2020-10-07 11:13:11.308987395 +0200 CEST m=+18.027368657]
I1007 11:13:11.310800  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [beta] [60s] [2020-10-07 11:13:11.31071294 +0200 CEST m=+18.029094199]
I1007 11:13:11.361233  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [default] [60s] [2020-10-07 11:13:11.361146623 +0200 CEST m=+18.079527912]
I1007 11:13:11.361378  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [electronic-shop] [60s] [2020-10-07 11:13:11.361313723 +0200 CEST m=+18.079694986]
I1007 11:13:11.361630  479093 client.go:92] TODELETE GetAllRequestRates [default] [60s] [2020-10-07 11:13:11.36106688 +0200 CEST m=+18.079448146]
I1007 11:13:11.361644  479093 client.go:92] TODELETE GetAllRequestRates [default] [60s] [2020-10-07 11:13:11.361170217 +0200 CEST m=+18.079551492]
I1007 11:13:11.361706  479093 client.go:92] TODELETE GetAllRequestRates [electronic-shop] [60s] [2020-10-07 11:13:11.361363865 +0200 CEST m=+18.079745125]
I1007 11:13:11.361708  479093 client.go:92] TODELETE GetAllRequestRates [electronic-shop] [60s] [2020-10-07 11:13:11.361352824 +0200 CEST m=+18.079734080]
I1007 11:13:11.389758  479093 client.go:92] TODELETE GetAllRequestRates [travel-agency] [60s] [2020-10-07 11:13:11.389299985 +0200 CEST m=+18.107681259]
I1007 11:13:11.389770  479093 client.go:92] TODELETE GetAllRequestRates [travel-agency] [60s] [2020-10-07 11:13:11.389365006 +0200 CEST m=+18.107746282]
I1007 11:13:11.389521  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [travel-agency] [60s] [2020-10-07 11:13:11.389454636 +0200 CEST m=+18.107835903]
I1007 11:13:11.392036  479093 client.go:92] TODELETE GetAllRequestRates [travel-portal] [60s] [2020-10-07 11:13:11.391732937 +0200 CEST m=+18.110114201]
I1007 11:13:16.609988  479093 client.go:92] TODELETE GetAllRequestRates [alpha] [60s] [2020-10-07 11:13:16.60932273 +0200 CEST m=+23.327703993]
I1007 11:13:16.610386  479093 client.go:92] TODELETE GetAllRequestRates [alpha] [60s] [2020-10-07 11:13:16.609424201 +0200 CEST m=+23.327805474]
I1007 11:13:16.623600  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [beta] [60s] [2020-10-07 11:13:16.623526782 +0200 CEST m=+23.341908056]
I1007 11:13:16.623692  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [alpha] [60s] [2020-10-07 11:13:16.623619778 +0200 CEST m=+23.342001043]
I1007 11:13:16.624141  479093 client.go:92] TODELETE GetAllRequestRates [beta] [60s] [2020-10-07 11:13:16.62371277 +0200 CEST m=+23.342094026]
I1007 11:13:16.624161  479093 client.go:92] TODELETE GetAllReI1007 11:13:16.667399  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [travel-agency] [60s] [2020-10-07 11:13:16.66730192 +0200 CEST m=+23.385683212]
questRates [beta] [60s] [2020-10-07 11:13:16.623459497 +0200 CEST m=+23.341840763]
I1007 11:13:16.632748  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [default] [60s] [2020-10-07 11:13:16.632640054 +0200 CEST m=+23.351021313]
I1007 11:13:16.632900  479093 client.go:92] TODELETE GetAllRequestRates [default] [60s] [2020-10-07 11:13:16.631428096 +0200 CEST m=+23.349809371]
I1007 11:13:16.645145  479093 client.go:92] TODELETE GetAllRequestRates [electronic-shop] [60s] [2020-10-07 11:13:16.644688812 +0200 CEST m=+23.363070068]
I1007 11:13:16.645155  479093 client.go:92] TODELETE GetAllRequestRates [default] [60s] [2020-10-07 11:13:16.644680999 +0200 CEST m=+23.363062257]
I1007 11:13:16.652108  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [electronic-shop] [60s] [2020-10-07 11:13:16.6520116 +0200 CEST m=+23.370392863]
I1007 11:13:16.652438  479093 client.go:92] TODELETE GetAllRequestRates [electronic-shop] [60s] [2020-10-07 11:13:16.651934386 +0200 CEST m=+23.370315659]
I1007 11:13:16.657551  479093 client.go:92] TODELETE GetAllRequestRates [travel-agency] [60s] [2020-10-07 11:13:16.657182038 +0200 CEST m=+23.375563299]
I1007 11:13:16.667399  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [travel-agency] [60s] [2020-10-07 11:13:16.66730192 +0200 CEST m=+23.385683212]
I1007 11:13:16.670963  479093 client.go:92] TODELETE GetAllRequestRates [travel-portal] [60s] [2020-10-07 11:13:16.670706295 +0200 CEST m=+23.389087558]
I1007 11:13:16.671061  479093 client.go:92] TODELETE GetAllRequestRates [travel-agency] [60s] [2020-10-07 11:13:16.670638042 +0200 CEST m=+23.389019320]
```

In some cases there are some repeated queries:

```
I1007 11:13:11.305598  479093 client.go:92] TODELETE GetAllRequestRates [alpha] [60s] [2020-10-07 11:13:11.305107932 +0200 CEST m=+18.023489238]
I1007 11:13:11.305641  479093 client.go:92] TODELETE GetAllRequestRates [alpha] [60s] [2020-10-07 11:13:11.305021299 +0200 CEST m=+18.023402559]

I1007 11:13:11.361706  479093 client.go:92] TODELETE GetAllRequestRates [electronic-shop] [60s] [2020-10-07 11:13:11.361363865 +0200 CEST m=+18.079745125]
I1007 11:13:11.361708  479093 client.go:92] TODELETE GetAllRequestRates [electronic-shop] [60s] [2020-10-07 11:13:11.361352824 +0200 CEST m=+18.079734080]

I1007 11:13:11.389758  479093 client.go:92] TODELETE GetAllRequestRates [travel-agency] [60s] [2020-10-07 11:13:11.389299985 +0200 CEST m=+18.107681259]
I1007 11:13:11.389770  479093 client.go:92] TODELETE GetAllRequestRates [travel-agency] [60s] [2020-10-07 11:13:11.389365006 +0200 CEST m=+18.107746282]
```

And in others it opens the discussion if in terms of business it makes sense to have two queries with due a short interval:

```
I1007 11:13:11.389521  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [travel-agency] [60s] [2020-10-07 11:13:11.389454636 +0200 CEST m=+18.107835903]
I1007 11:13:16.667399  479093 client.go:102] TODELETE GetNamespaceServicesRequestRates [travel-agency] [60s] [2020-10-07 11:13:16.66730192 +0200 CEST m=+23.385683212]
```

In this last one two queries with a difference of 5 seconds.

So, this opens to study to cache some results first in a short-live manner as we similar we do for namespaces or proxy-status.

That would address some preliminary concerns without changing the architecture or the high level design of all pieces we have in place.

Some seems obvious to me: repeated queries doesn't make sense, those can be easily added in memory, and probably more tweaks can be done if a result can be valid in a short-time period (<1m) something like that, that would accelerate specially scenarios when several queries are collected for a single operation (i.e. graph or list generation).

**Comment by jshaughn** [comment]
@lucasponce I am OK with releasing this feature and getting any upstream feedback that  it may generate.  As you commented in the PR, despite ech graph refresh maybe taking longer, it may not be noticeable to the user and the extra time results in a smoother update and the ability to find/hide on health.  I think the remaining question is only which impl to go with, at least initially.  The approach in https://github.com/kiali/kiali-ui/pull/1957 maybe fetches less health overall [1] but fetches it all only after the graph-gen returns.  The approach in https://github.com/kiali/kiali-ui/pull/1964 maybe fetches more health overall [2] but performs some of the health fetch concurrently with graph-gen.

[1] It fetches only namespace-level health for only the types present in the graph.  For example, if 1 workload node for ns=foo, we fetch all workload health for foo.  All fetching is post-graph-gen because the health fetched depends on the graph nodes.

In my tests using 2 namespaces (istio-system and bookinfo) the average fetch is about 800ms.  Graph-gen/health is about 60%/40% of total time.

[2] It fetches namespace-health for all types for all selected namespaces (note app health is only fetched for app graph-types).  It is possible that the graph may not have nodes for the fetched health, but if we assume that most graphs have traffic for all services, then it's not a bad approach.  The fetch can be done concurrently with graph-gen.  Post graph-gen we then need to fetch individual health for accessible "outsider" nodes.

In my tests using 2 namespaces (istio-system and bookinfo) the average fetch is about 750ms.  Graph-gen/health is about equal time, plus some outsider health at the end.  Graph-gen is alctually slowed, probably because of prometheus contention given health happening concurrently.

The second approach seems a bit faster but not much, in my tests.  I would like another set of data points.  The PRs have console logging that will dump some basic perf stats.

**Comment by israel-hdez** [comment]
I was seeing the code of the "concurrent" approach. Quite the same. If my eyes catched correctly, it's only the GraphDataSource that changed.

Well, I'm not sure there is a better approach. I would take [1], but I think I would have a little biased opinion. My reasons: I prefer to keep servers/network with the least load (which it's potentially increased in [2], even if it's just a little); and because by not increasing the load, it's cheaper if running in the cloud :)
